NIFI-1022 Added Tachyon/Alluxio processors

pvillard31 opened the pull request:
- ListAlluxio processor
- PutAlluxio processor
- GetAlluxio processor

apiri commented:
Starting review.

For anyone who was interested in checking it out, I was able to locate a Docker image that seems quite nice.

You can run it locally via:
`docker run -it -p 19999:19999 -p 30000:30000 --name alluxio harisekhon/alluxio`

apiri commented on nifi-nar-bundles/nifi-alluxio-bundle/nifi-alluxio-processors/src/main/java/org/apache/nifi/processors/alluxio/AbstractAlluxioProcessor.java:
```diff
@@ -0,0 +1,113 @@
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * AbstractCassandraProcessor is a base class for Alluxio processors and contains logic and variables common to most
```
Cassandra -> Alluxio

pvillard31 replied:
good catch, done

apiri commented on nifi-nar-bundles/nifi-alluxio-bundle/nifi-alluxio-processors/src/main/java/org/apache/nifi/processors/alluxio/AbstractAlluxioProcessor.java:
```diff
@@ -0,0 +1,113 @@
+    protected final AtomicReference<FileSystem> fileSystem = new AtomicReference<>(null);
+
+    protected void createFileSystem(ProcessContext context) {
+        System.setProperty("alluxio.master.hostname", context.getProperty(MASTER_HOSTNAME).getValue());
```
Is there no way around these system properties?

pvillard31 replied:
I didn't find another way to instantiate it... but I agree this is not ideal.

apiri commented on nifi-nar-bundles/nifi-alluxio-bundle/nifi-alluxio-processors/src/main/java/org/apache/nifi/processors/alluxio/GetAlluxio.java:
```diff
@@ -0,0 +1,184 @@
+            .build();
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("In case of failure, flow files will be routed to this relationship")
```
minor:  Could we standardize the use of flow file/FlowFile throughout these properties?

apiri commented on nifi-nar-bundles/pom.xml:
```diff
@@ -59,6 +59,7 @@
         <module>nifi-jms-bundle</module>
         <module>nifi-cassandra-bundle</module>
         <module>nifi-spring-bundle</module>
+        <module>nifi-alluxio-bundle</module>
```
Will also need to provide the alluxio-nar in the nifi-assembly pom and as a managed dependency in the root pom.

apiri commented on nifi-nar-bundles/nifi-alluxio-bundle/nifi-alluxio-processors/src/main/java/org/apache/nifi/processors/alluxio/AbstractAlluxioProcessor.java:
```diff
@@ -0,0 +1,113 @@
+    protected void createFileSystem(ProcessContext context) {
+        System.setProperty("alluxio.master.hostname", context.getProperty(MASTER_HOSTNAME).getValue());
+        System.setProperty("alluxio.master.port", context.getProperty(MASTER_PORT).getValue());
+        fileSystem.set(FileSystem.Factory.get());
```
This doesn't seem to update regardless of changing a property.  I accidentally used the wrong port when testing and despite changing in the properties, stopping/starting or disabling/enabling, the backing Alluxio client seems stuck on the initial port.

apiri commented on nifi-nar-bundles/nifi-alluxio-bundle/nifi-alluxio-processors/src/main/java/org/apache/nifi/processors/alluxio/GetAlluxio.java:
```diff
@@ -0,0 +1,184 @@
+            flowFile = session.write(flowFile, new OutputStreamCallback() {
+                @Override
+                public void process(final OutputStream out) throws IOException {
+                    IOUtils.copy(toCopy, out);
```
This is throwing errors in some cases and not bubbling out to the wrapping catch so there is no indication of issues.  One particular instance was:

> 016-06-19 01:12:52,512 ERROR [Timer-Driven Process Thread-8] o.a.nifi.processors.alluxio.PutAlluxio PutAlluxio[id=37744f8c-57fe-48c9-ac0b-e3fd255efc4b] An error occurred while writing file /nifi-630c823e-7492-4539-bf14-632dfa4b85e2; transferring to 'failure': java.lang.RuntimeException: No available Alluxio worker found
2016-06-19 01:12:52,513 ERROR [Timer-Driven Process Thread-8] o.a.nifi.processors.alluxio.PutAlluxio
java.lang.RuntimeException: No available Alluxio worker found
	at alluxio.client.block.AlluxioBlockStore.getOutStream(AlluxioBlockStore.java:174) ~[alluxio-core-client-1.0.1.jar:na]
	at alluxio.client.file.FileOutStream.getNextBlock(FileOutStream.java:266) ~[alluxio-core-client-1.0.1.jar:na]
	at alluxio.client.file.FileOutStream.write(FileOutStream.java:231) ~[alluxio-core-client-1.0.1.jar:na]
	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1793) ~[commons-io-2.4.jar:2.4]
	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1769) ~[commons-io-2.4.jar:2.4]
	at org.apache.commons.io.IOUtils.copy(IOUtils.java:1744) ~[commons-io-2.4.jar:2.4]

apiri replied:
It seems like the Alluxio client logger may be interfering with NiFi's logger and is causing bulletins to not work appropriately.

pvillard31 replied:
What do you mean/suggest?

apiri replied:
I'll have to scope this out again.  After the fact I realized that it may have just been the bulletin bug that was on master that was causing bulletins to work.  When I next get the opportunity, I'll see if that's still the case.

For some additional context, in the logs, which I now realize I failed to capture, there was mention of the Alluxio logger.  I was curious if that was somehow interfering with NiFi's logger and preventing bulletins.  I'll let you know if I can recreate.

apiri commented on nifi-nar-bundles/nifi-alluxio-bundle/nifi-alluxio-processors/src/main/java/org/apache/nifi/processors/alluxio/AbstractAlluxioProcessor.java:
```diff
@@ -0,0 +1,113 @@
+    public static final PropertyDescriptor URI = new PropertyDescriptor.Builder()
+            .name("alluxio-uri")
+            .displayName("URI")
+            .description("Alluxio URI to use. Example: /path")
```
This is a little unclear to me in the context of a file.  I had specified a static path, but then received errors about the specified path already existing.  Accordingly, my assumption was then that this is path + file like an S3 bucket, however I didn't seem to have significantly better luck in that scenario either.

pvillard31 replied:
I assumed the user would use expression language to define the file name with Get/Put. Something like /${filename} for Put, and /${alluxio_name} for Get. But if this is unclear, I certainly need to update descriptions. FYI, in the JIRA, I added one of the templates I used when testing the processors.

apiri replied:
Oh, good to know and that is super helpful.  Apologies for overlooking it as I sometimes neglect JIRA while in GitHub land.

apiri commented:
Hey @pvillard31.

Gave an initial scan over this code and played around a bit with Alluxio.  I didn't have much success interacting with the processors but have left some notes on some of the items I was confused with whilst using.

While I am thinking of it, could you get this rebased/updated to master?

I'll try to give it a fresh look when I have some spare moments to scope out the Alluxio docs a bit more and try to figure out what is going on.

pvillard31 commented:
Thanks for the review @apiri!
I still need to take into account some of your comments but I rebased the PR against master. While doing some additional testing, for a reason I can't explain yet, I was unable to use the ListProcessor:

````
2016-06-21 23:28:56,127 WARN [Timer-Driven Process Thread-9] o.a.n.c.t.ContinuallyRunProcessorTask Administratively Yielding ListAlluxio[id=42835c14-3896-47e0-95ae-ce8266e7c030] due to uncaught Exception: java.lang.NullPointerException: Name is null
2016-06-21 23:28:56,130 WARN [Timer-Driven Process Thread-9] o.a.n.c.t.ContinuallyRunProcessorTask 
java.lang.NullPointerException: Name is null
	at java.lang.Enum.valueOf(Enum.java:236) ~[na:1.8.0_77]
	at alluxio.exception.AlluxioExceptionType.valueOf(AlluxioExceptionType.java:19) ~[na:na]
	at alluxio.exception.AlluxioException.from(AlluxioException.java:69) ~[na:na]
	at alluxio.AbstractClient.retryRPC(AbstractClient.java:324) ~[na:na]
	at alluxio.client.file.FileSystemMasterClient.listStatus(FileSystemMasterClient.java:271) ~[na:na]
	at alluxio.client.file.BaseFileSystem.listStatus(BaseFileSystem.java:188) ~[na:na]
	at alluxio.client.file.BaseFileSystem.listStatus(BaseFileSystem.java:179) ~[na:na]
	at org.apache.nifi.processors.alluxio.ListAlluxio.onTrigger(ListAlluxio.java:91) ~[na:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1077) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:136) [nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:47) [nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:123) [nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_77]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_77]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_77]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_77]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_77]
````

I'll try to dig more into this tomorrow. Otherwise, a guy from Alluxio reached me to propose his help if needed on this PR, I guess I can ask him to have a look and make some comments.

pvillard31 commented:
Hey guys,

I've done some rework on this PR. After some discussions with Alluxio folks, it seems that there are some limitations at the moment (v1.3.0) and I've made some modifications to take into account those limitations.

### System properties
At the moment, it seems it is not possible to initialize the Alluxio client inside NiFi without using system properties. And in any case, multi Alluxio clusters handling is not supported at the moment.

### Initialization
Once the client is initialized, it cannot be unchanged unless NiFi is restarted. Even if provided IP/port are incorrect, the client initialization will not do any check. To avoid mistakes, I've added a simple check to ensure that something is listening at IP/port before initializing the client.
*Improvement on Alluxio's side is tracked here: https://alluxio.atlassian.net/browse/ALLUXIO-2120*

I've updated processors description to note it as **experimental** and let the users know what are the current limitations. Hopefully, this won't be a show stopper to get this first version into NiFi and get feedbacks from the community.

apiri commented:
@pvillard31 I saw that the referenced ticket seems to now be resolved.  Not sure if this was still of interest, but if so would be happy to revisit upon updating.

pvillard31 commented:
Hey @apiri, yes Alluxio guys informed me about this, I'll try to update the PR asap. Will let you know when it's done.

apiri commented:
no rush!  was just trying to revisit items in the queue

joewitt commented:
@pvillard31 do you still think we can get this across the finish line?

pvillard31 commented:
Hey @joewitt, I will update the PR asap this week.

pvillard31 commented:
@apiri @joewitt 
I updated the PR with the following changes:
- Alluxio version changed to 1.4.0 to include the expected improvement (https://github.com/Alluxio/alluxio/pull/4422) and removed the test I was doing before
- changed the GetAlluxio to a FetchAlluxio
- changed the ListAlluxio to include a state to store the timestamp of the most recent file listed by the processor
- updated the unit tests to match the changes

pvillard31 commented:
and... almost forgot to link a template to play with the processors:
https://gist.github.com/pvillard31/2bf996adefaf1709339451d71c343ac1

joewitt commented:
i dont know anything about the target systems so I probably won't be a great reviewer.  If nobody else is in a good position to help though I'll give it a whirl.

joewitt commented:
@jdye64 any chance you could have some cycles to review this as the originator of the JIRA?

trixpan commented:
@pvillard31 not sure if it is the case but have you by any chance tried to use the Hadoop alluxio interfaces with puthdfs and co instead of a new processor? 

e.g


https://www.alluxio.com/docs/community/1.3/en/Running-Hadoop-MapReduce-on-Alluxio.html

trixpan commented:
And by last comment I naturally refer to use a customised core-site.xml that includes:

```
<property>
  <name>fs.alluxio.impl</name>
  <value>alluxio.hadoop.FileSystem</value>
  <description>The Alluxio FileSystem (Hadoop 1.x and 2.x)</description>
</property>
<property>
  <name>fs.alluxio-ft.impl</name>
  <value>alluxio.hadoop.FaultTolerantFileSystem</value>
  <description>The Alluxio FileSystem (Hadoop 1.x and 2.x) with fault tolerant support</description>
</property>
<property>
  <name>fs.AbstractFileSystem.alluxio.impl</name>
  <value>alluxio.hadoop.AlluxioFileSystem</value>
  <description>The Alluxio AbstractFileSystem (Hadoop 2.x)</description>
</property>
```

And the class_path technique described here https://community.hortonworks.com/articles/71916/connecting-to-azure-data-lake-from-a-nifi-dataflow.html

pvillard31 commented:
Hey @trixpan, I didn't try but maybe that would work, I honestly don't know. However, Alluxio can run on top of a lot of storage solutions, so it makes sense to have a dedicated processor. Anyway, I see there is a conflict, I'll need to rebase the PR, will do tomorrow.

trixpan commented:
@pvillard31 that connection strategy is agnostic to the underlying filesystem supporting alluxio. What it does is to allow a hadoop compatible application to access alluxio without further coding and let alluxio  deal with the underlying stirage strategies

joewitt commented:
@pvillard31 this poor PR has been hanging out for quite a while.  Shall we try again?  Can you update?
